Refactor login preference handling into its own class

bguertin opened the pull request:
So, I don't like how so many classes are touching the same preference file ad-hoc. No behaviour change is intended by this cleanup, except for bug-fixes.

- Updated every reference to the "LOGIN" preference file to go through a single class, `LoginPrefs`.
- Removed the login-specific methods from `PrefManager`, since they are now all handled by `LoginPrefs`
- Fixed a bug where logging in via Facebook or Google would be recorded to Segment.IO as logging in via password.
- Removed code (that never ran) that seems to have been trying to force users to log out if they logged in via Facebook/Google. I don't know what it's purpose was, but it was not working anyway.
- Deleting that code also fixes a bug where the last activity instance was forced to remain in memory.
- Also fixed https://openedx.atlassian.net/browse/MA-2489
- Moved login-related methods out of `ServiceManager` and directly into `LoginService`. Deleted the old deprecated implementations from `Api`, in favor the Retrofit implementation.
- Updated unit/feature tests to use `LoginApi` instead of the deprecated `Api` methods
- Deleted unused `TEST_ACCOUNT` config property

TODO:
 - [x] Manually test Google login / registration
 - [x] Manually test Facebook login / registration
 - [x] Manually test Email login / registration
 - [x] Manually Test OAuth token refresh

BenjiLee commented:
Marking as WIP until this is ready for reviews :P

bguertin commented:
Getting this error from Google sign in, so not ready yet.

```
06-09 12:45:00.491 23154-23500/org.edx.mobile W/DefaultRequestDirector: Authentication error: Unable to respond to any of these challenges: {bearer=WWW-Authenticate: Bearer realm="api"}
```

bguertin commented:
Ok, I've tested everything I could think of and everything is working as before (or better).

Ready for review.

BenjiLee commented:
CI is angry? Rabase?

BenjiLee commented:
@bguertin Any strategy to reviewing this?

bguertin commented:
Rebased. Hopefully CI works this time.

bguertin commented:
> Any strategy to reviewing this?

Ignore whitespace: https://github.com/edx/edx-app-android/pull/728/files?w=1

I would focus attention on `LoginApi` which is where all the magic happens. All the other changes are refactoring to call methods in that class instead of doing the preference editing ad-hoc.

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/authentication/LoginAPI.java:
```diff
@@ -1,38 +1,173 @@
 import retrofit.RestAdapter;
 
-/**
- * Created by cleeedx on 4/12/16.
```
Nouuuuu

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/authentication/LoginAPI.java:
```diff
@@ -1,38 +1,173 @@
 
     @Inject
-    public LoginAPI(@NonNull RestAdapter restAdapter) {
+    public LoginAPI(@NonNull RestAdapter restAdapter, @NonNull Config config, @NonNull LoginPrefs loginPrefs, @NonNull ServiceManager serviceManager, @NonNull ISegment segment, @NonNull NotificationDelegate notificationDelegate) {
```
Way too many characters on one line!!!

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/authentication/LoginAPI.java:
```diff
@@ -1,38 +1,173 @@
+            finishLogIn(response, LoginPrefs.AuthBackend.PASSWORD, email.trim());
+            return response;
+        } catch (HttpResponseStatusException ex) {
+            if (ex.getStatusCode() >= 400 && ex.getStatusCode() < 500) {
```
Are we guaranteed a non-null value for .getStatusCode?

bguertin replied:
`getStatusCode` returns `int` which cannot be null.

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/authentication/LoginAPI.java:
```diff
@@ -1,38 +1,173 @@
+
+    @NonNull
+    public AuthResponse logInUsingFacebook(String accessToken) throws Exception {
+        final AuthResponse response = serviceManager.loginByFacebook(accessToken);
```
Can we delete the deprecated loginByFacebook? Same for google.

bguertin replied:
Hmm, you mean move the calls into `LoginService` and out of ServiceManager? that's a good idea

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/authentication/LoginAPI.java:
```diff
@@ -1,38 +1,173 @@
+    }
+
+    private void finishSocialLogIn(@NonNull AuthResponse response, @NonNull LoginPrefs.AuthBackend authBackend) throws Exception {
+        if (response.error != null && response.error.equals("401")) {
```
This check seems a bit specific. Looks like this is what the old code did.

bguertin replied:
Yeah I just moved the check, There were no comments explaining it's reasoning :(

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/http/Api.java:
```diff
@@ -153,39 +147,17 @@ public ProfileModel getProfile() throws Exception {
-        if (json == null) {
-            return null;
+        final HttpManager.HttpResult result = http.get(urlWithAppendedParams, getAuthHeaders());
+        if (result.statusCode < 200 || result.statusCode >= 300) {
```
This is also a weird comparison.

bguertin replied:
It's checking for a successful status code. Something which Retrofit does automatically. Perhaps I should also just delete this method and use the retrofit one? I think so...

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/http/OauthHeaderRequestInterceptor.java:
```diff
@@ -1,44 +1,44 @@
+            builder.addHeader("Authorization", token);
         } else {
-            builder.addHeader("Authorization", String.format("%s %s", auth.token_type, auth.access_token));
+            logger.warn("Token cannot be null when AUTH_JSON is also null, something is WRONG!");
```
MAybe this is out of scope:

I always see this in the log which makes it seem like there is always something wrong. Is this supposed to always come up as a warning?

BenjiLee replied:
I also see this warning in the Course Unit fragment. Maybe someone copy pastaed unintentionally to the login class.

bguertin replied:
I don't know what the warning was put there for. I'll just remove it I guess.

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/task/RegisterTask.java:
```diff
@@ -3,75 +3,38 @@
-                if (auth.isSuccess()) {
-                    logger.debug("login succeeded after email registration");
-                }
+                return loginAPI.registerUsingEmail(parameters);
```
This looks SOOOO MUCH BETTER!

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/view/LaunchActivity.java:
```diff
@@ -1,93 +1,76 @@
     protected boolean createOptionsMenu(Menu menu) {
-        // Launch screen doesn't have any menu
-        return true;
+        return false; // Disable menu inherited from BaseFragmentActivity
```
Why does this even have an override if there is no menu?

bguertin replied:
Because our base activity class `BaseFragmentActivity` has an option menu by default and we want to suppress it.

On my long list of things to clean up if I have time: https://openedx.atlassian.net/browse/MA-2093?filter=-1

BenjiLee commented:
This more or less LGTM. I played around with a bit. I didn't get to manually test the google login and registration yet. Early 👍  from me, I will be going PTO after today. I'm sure if there is anything alarming that happens, you would be able to handle it :)

bguertin commented:
Hate to increase scope, but made some additional much needed cleanup:
- Moved login-related methods out of `ServiceManager` and directly into `LoginService`. Deleted the old deprecated implementations from `Api`, in favor the new Retrofit implementation.
- Updated unit and feature tests to use `LoginApi`/`LoginPrefs` instead of the deprecated `ServiceManager`/`Api` methods

bguertin commented:
I also manually retested all the log in / sign-up cases that I'm aware of. Everything is still working great as far as I can tell.

I think we can merge this right *after* tagging the next release, and continue to exercise it during the next dev cycle.

BenjiLee commented:
SGTM. Will be able to do a more thorough pass.

bguertin commented:
Fixed the automated tests, so this is good to merge after review. Please let me know if you have questions, since I know many lines were touched (didn't expect this to be such a large task).

BenjiLee commented:
@bguertin We should get this ready to merge since it's now a blocker for the 2.5.0 release.

bguertin commented:
Yup. I personally think this is release-worthy. I suggested delaying so as to not slow down the release :laughing: . Didn't know these were major issues.

BenjiLee commented:
Update: We're going to wait until 2.5.0 is released :/ QA for the current 2.5.0 will be much simpler. I'll still give this a review :)

BenjiLee commented:
Smoke tests:
-[] fefe

BenjiLee commented:
👍
